components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginStorageDelegate.kt

```diff
@@ -0,0 +1,146 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.sync.logins
+
+import androidx.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import mozilla.components.concept.storage.Login
+import mozilla.components.concept.storage.LoginStorageDelegate
+import mozilla.components.lib.dataprotect.SecureAbove22Preferences
+
+/**
+ * A type of persistence operation, either 'create' or 'update'.
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+enum class Operation { CREATE, UPDATE }
+
+internal const val PASSWORDS_KEY = "passwords"
+
+/**
+ * [LoginStorage.Delegate] implementation.
+ *
+ * An abstraction that handles the persistence and retrieval of [LoginEntry]s so that Gecko doesn't
+ * have to.
+ *
+ * In order to use this class, attach it to the active [GeckoRuntime] as its `loginStorageDelegate`.
+ * It is not designed to work with other engines.
+ */
+class DefaultLoginStorageDelegate(
+    private val loginStorage: AsyncLoginsStorage,
+    keyStore: SecureAbove22Preferences,
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
+) : LoginStorageDelegate {
+
+    private val password = { scope.async { keyStore.getString(PASSWORDS_KEY)!! } }
+
+    override fun onLoginUsed(login: Login) {
+        val guid = login.guid
+        if (guid == null || guid.isEmpty()) return
```

Comment:
Maybe worth leaving a comment saying that you're essentially checking if we actually have this `login` saved, and we use presence of a `guid` on the object as a proxy for "we have this login in the db" (as opposed to looking it up, for example).

This is also assuming a few things about GV's behaviour, either:
- ordering of calls - `onLoginSave` will be called _before_ `onLoginUsed`
- or, `onLoginUsed` won't be called after `onLoginSave` at all - e.g. GV is assuming we'll mark login as "used" while saving/updating it
- or, ...

Can you document in a comment which behaviour you're assuming here?

Reply:
I believe this method is called when GV fills an existing login into a page, and it is not expected to be called during saving at all.  The API hasn't been hooked up yet so I wasn't able to verify that.